lib/trino-hive-formats/src/main/java/io/trino/hive/formats/line/json/JsonDeserializer.java

```diff
@@ -0,0 +1,838 @@
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static io.trino.hive.formats.HiveFormatUtils.createTimestampParser;
+import static io.trino.hive.formats.HiveFormatUtils.parseHiveDate;
+import static io.trino.hive.formats.HiveFormatUtils.writeDecimal;
+import static io.trino.plugin.base.type.TrinoTimestampEncoderFactory.createTimestampEncoder;
+import static io.trino.spi.type.BigintType.BIGINT;
+import static io.trino.spi.type.BooleanType.BOOLEAN;
+import static io.trino.spi.type.Chars.truncateToLengthAndTrimSpaces;
+import static io.trino.spi.type.DateType.DATE;
+import static io.trino.spi.type.Decimals.overflows;
+import static io.trino.spi.type.DoubleType.DOUBLE;
+import static io.trino.spi.type.IntegerType.INTEGER;
+import static io.trino.spi.type.RealType.REAL;
+import static io.trino.spi.type.RowType.field;
+import static io.trino.spi.type.SmallintType.SMALLINT;
+import static io.trino.spi.type.TinyintType.TINYINT;
+import static io.trino.spi.type.VarbinaryType.VARBINARY;
+import static io.trino.spi.type.Varchars.truncateToLength;
+import static java.lang.Float.floatToRawIntBits;
+import static java.lang.StrictMath.toIntExact;
+import static java.lang.String.format;
+import static java.nio.charset.StandardCharsets.UTF_8;
+import static org.joda.time.DateTimeZone.UTC;
+
+/**
+ * Deserializer that is bug for bug compatible with Hive JsonSerDe where possible. Known exceptions are:
+ * <ul>
+ *     <li>
+ *         When a scalar value is actually a json object, Hive will process the open curly bracket for
+ *         BOOLEAN, DECIMAL, CHAR, VARCHAR, and VARBINARY.  Then it continues processing field inside of
+ *         the json object as if they are part of the outer json object.  When the closing curly bracket
+ *         is encountered it pops a level, which can end parsing early.  This is clearly a bug resulting
+ *         in corrupted data, and instead we throw an exception.
+ *     </li>
+ *     <li>
+ *         Duplicate json object fields are supported, and like Hive. Hive parses each of these duplicate
+ *         values, but this code only process the last value.  This means if one of the duplicates is
+ *         invalid, Hive will fail, and this code will not.
+ *     </li>
+ * </ul>
+ */
+public class JsonDeserializer
+        implements LineDeserializer
+{
+    private final List<Type> types;
+    private final JsonFactory jsonFactory;
+    private final RowDecoder rowDecoder;
+
+    public JsonDeserializer(List<Column> columns, List<String> timestampFormats)
+    {
+        this.types = columns.stream()
+                .map(Column::type)
+                .collect(toImmutableList());
+
+        // Interning all encountered field names improves little compared to the per-factory canonical name cache and can cause native memory issues
+        jsonFactory = new JsonFactory();
+//              .disable(JsonParser.Feature.INTERN_FIELD_NAMES);
```

Comment:
We probably do want to disable this. In `JsonFunctions`, we disable `CANONICALIZE_FIELD_NAMES` which also disables interning. According to [this article](https://shipilev.net/jvm/anatomy-quarks/10-string-intern/) using string interning is slow and causes GC issues.